ward/_terminal.py

```diff
@@ -789,68 +798,149 @@ def get_source(self, err: TestFailure, test_result: TestResult) -> RenderableTyp
+                ("item ", "pass.textonly"),
+                *self.of_type(err.lhs),
+            )
+            rhs_msg = Text.assemble(
+                (
+                    "was not " if err.operator is Comparison.In else "was ",
+                    "bold default",
+                ),
+                ("found in the ", "default"),
+                ("container ", "fail.textonly"),
+                *self.of_type(err.rhs),
+            )
+        elif err.operator in IS_COMPARISONS:
+            lhs_msg = Text.assemble(
+                ("The ", "default"),
+                ("LHS ", "pass.textonly"),
+                *self.of_type(err.lhs),
+                (" with ", "default"),
+                ("id ", "default"),
+                (f"{id(err.lhs)}", "bold default"),
+            )
+            rhs_msg = Text.assemble(
+                (
+                    "was not " if err.operator is Comparison.Is else "was ",
+                    "bold default",
+                ),
+                ("the ", "default"),
+                ("RHS ", "fail.textonly"),
+                *self.of_type(err.rhs),
+                (" with ", "default"),
+                ("id ", "default"),
+                (f"{id(err.rhs)}", "bold default"),
+            )
+        else:
+            raise Exception("unknown operator")
 
-    def get_pretty_failure_for_in(self, err: TestFailure) -> RenderableType:
-        lhs_msg = Text.assemble(
-            ("The ", "default"),
-            ("item ", "pass.textonly"),
-            *self.of_type(err.lhs),
-        )
         lhs = Panel(
             Pretty(err.lhs),
             title=lhs_msg,
             title_align="left",
             border_style="pass.textonly",
             padding=1,
-        )
-
-        rhs_msg = Text.assemble(
-            ("was not " if err.operator is Comparison.In else "was ", "bold default"),
-            ("found in the ", "default"),
-            ("container ", "fail.textonly"),
-            *self.of_type(err.rhs),
+            expand=True,
         )
         rhs = Panel(
             Pretty(err.rhs),
             title=rhs_msg,
             title_align="left",
             border_style="fail.textonly",
             padding=1,
+            expand=True,
         )
 
-        return Padding(RenderGroup(lhs, rhs), pad=(0, 0, 1, 2))
+        return Columns(
+            [lhs, rhs],
+            expand=True,
+            padding=0,
+        )
 
     def of_type(self, obj: object) -> Iterator[Tuple[str, str]]:
         yield "(of type ", "default"
         yield type(obj).__name__, "bold default"
         yield ")", "default"
 
+    def get_diff(self, err: TestFailure) -> Optional[RenderableType]:
+        diff = None
+
+        if err.operator in EQUALITY_COMPARISONS:
+            diff = Diff(
+                err.lhs,
+                err.rhs,
+                width=self.terminal_size.width - 24,
+                show_symbols=self.show_diff_symbols,
+            )
```

Comment:
I think this is where custom diffs would go - I'm imagining eventually pulling this out into a hook mechanism, where the hook gets passed the error and can decide whether you can produce the diff for it. We'd have to be careful with the ordering, to make sure that more-specific handlers can get themselves registered first, and more-general ones last.

Reply:
Here's a rough sketch of how I could see that working:

We'd update the code you highlighted above to call a hook which returns a `Diff` or some other renderable:

```python
# call the hook
diffs_from_plugins = plugins.hook.get_diff(config, err, term_width)

# display the first result from the hook
console.print(diff[0])
```

(We define the hook by defining a new spec)...

```python
@spec(firstresult=True)  # only call the first plugin that implements this hook
def get_diff(config, err, term_width) -> Optional[ConsoleRenderable]:
    pass
```

and Ward would offer a default implementation of that hook that just returns the `Diff` object we currently return:

```python
@hook
def get_diff(config, err, term_width) -> Optional[ConsoleRenderable]:
    return Diff(...)
```

A plugin would "override" that hook (need to be careful with registration order - if only the first result is used since `firstresult=True`, then we need to make sure the plugins are registered **after** the default implementations at the bottom of `hooks.py` (since pluggy is LIFO):

```python
# Inside the plugin...

class MyFancyDiff:
    def __rich_console__(...):
        ...

@hook
def get_diff(config, err, term_width) -> Optional[ConsoleRenderable]:
    return MyFancyDiff(...)
```

Reply:
On further thought, we'd probably want to make this entire method a hook, since the panel text "LHS vs RHS shown below" might not make sense for a custom diff output, and so plugins may wish to change that too.